packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts

```diff
@@ -84,21 +94,14 @@ export class RemoteMessageProcessor {
 	}
 }
 
-const copy = (remoteMessage: ISequencedDocumentMessage): ISequencedDocumentMessage => {
-	// Do shallow copy of message, as the processing flow will modify it.
-	// There might be multiple container instances receiving same message
-	// We do not need to make deep copy, as each layer will just replace message.content itself,
-	// but would not modify contents details
-	const message = { ...remoteMessage };
-
+/** Takes an incoming message and if the contents is a string, JSON.parse's it in place */
+const ensureContentsDeserialized = (mutableMessage: ISequencedDocumentMessage): void => {
```

Comment:
Not necessarily for this PR, but I wonder if we can use `readonly` to help auto-document when this is mutable and when it isn't. For example, make all the fields on ISequencedDocumentMessage `readonly`. Then, this area of the code would operate on the copy which drops the `readonly` from the contents field.

Reply:
Well, on second thought not sure how that helps.  We actually _don't_ want `contents` to be readonly because we want to replace it. Rather we want its properties to be readonly.  At runtime we could do `Object.freeze` to guarantee it I guess? But not sure how to type it.  Btw it's already updated to `contents: unknown` in the latest version (not yet consumed here).

@jason-ha, thoughts?

Reply:
Right - I just mean it should normally have all readonly fields (including `contents`), but that this is the special area that `contents` would become temporarily mutable. In general, it is good practice to mark all fields of an interface readonly, except in the places where they are expected to be mutated; we clearly don't do that for ISequencedDocumentMessage though.

Reply:
`readonly` is good. Commonplace use of it will help make it clear when things are mutable. There are could be three typings for three cases for these structures.
1. all fields readonly because this is incoming and others may look at it
2. still processing where some fields are mutable (you have to make a shallow copy to make things mutable)
3. ready to apply where fields are readonly

I think this is more or less what peoples were thinking. Just writing it out a little differently to distinguish we avoid making something mutable once it is readonly. (You want a new object for that.)